### DISCUSSION TITLE
LinkError thrown in browser when trying to use web browser bevy game

### DISCUSSION THREAD

**Discussion by kpence** [bevyengine/bevy#4888] [Q&A] [answered]
For some reason, I can't use bevy to make web browser games without running into this issue. This very minimal example will reproduce it. I get "LinkError" messages thrown in the browser's console, and a white screen, and my game doesn't get run.

The error that gets thrown in the browser is (although in the past, when I had this same kind of issue, I would get other similar LinkError messages):
```
Uncaught (in promise) LinkError: import object field '__wbindgen_object_drop_ref' is not a Function [game-test:26:19](https://site/game/)
    run https://site/game/:26
    AsyncFunctionThrow self-hosted:704
    (Async: async)
    <anonymous> https://site/game/:57
    InnerModuleEvaluation self-hosted:2393
    evaluation self-hosted:2354
```

I should expect to get a message like "panic: foobar", but I don't.

The command I'm running is: `cargo run  --example game --target wasm32-unknown-unknown`

Here's all the files for reproducing this:

.cargo/config.toml:
```
[target.wasm32-unknown-unknown]
runner = "wasm-server-runner"
```

Cargo.toml:
```
[package]
name = "bug"
version = "0.1.0"
edition = "2021"

[lib]
crate-type = ["cdylib", "rlib"]

[dependencies]
console_error_panic_hook = "0.1"
bevy = { version = "0.7", default-features = false, feature = ["dynamic", "bevy_core_pipeline" ] }
wasm-bindgen = "*"

[[example]]
name = "game"
path = "examples/game/main.rs"
```

examples/game/main.rs:
```rs
use bug::run;

fn main() {
    run();
}
```

src/lib.rs:
```rs
use bevy::prelude::*;
use wasm_bindgen::prelude::*;

#[wasm_bindgen]
pub fn run() {
    #[cfg(target_arch = "wasm32")] 
    console_error_panic_hook::set_once();
    App::new()
        .add_startup_system(||{panic!("foobar");})
        .run();
}
```

Is anyone else running to the same issue with this example?

**Comment by jakobhellermann** [comment, marked as answer]
What version of `wasm-bindgen` do you have in your project (`cargo tree --invert wasm-bindgen --target wasm32-unknown-unknown --depth 0`, should be `0.2.80`)? What version of `wasm-server-runner` do you have (`cargo install --list | grep wasm-server-runner`, should be >`0.2.3` I think)?

I have seen similar errors when there is a version mismatch between the wasm-bindgen version and the one used in wasm-server-runner.
Perhaps `cargo install wasm-server-runner` fixes the issue.

**Reply by jakobhellermann** [reply]
Starting with `wasm-server-runner 0.2.4` you can also run `--version` to find out:
```
λ wasm-server-runner --version
wasm-server-runner: 0.2.4,
wasm-bindgen: 0.2.80
```

**Reply by kpence** [reply]
> `cargo tree --invert wasm-bindgen --target wasm32-unknown-unknown --depth 0`
I get:
`v0.2.79`

wasm-server-runner is:
`v0.2.1`

I'll see what happens when I update wasm-server-runner and explicitly set the version for wasm-bindgen

**Reply by kpence** [reply]
Updating wasm-server-runner and wasm-bindgen fixed the issue, thanks! I'll see if it works on my actual game now

**Reply by kpence** [reply]
So, the example I posted in the original post passed after making these changes, but my actual game is having the same kind of issue and throwing this error in the browser's console:

`Uncaught (in promise) LinkError: import object field '__wbindgen_object_clone_ref' is not a Function`

**Reply by mockersf** [reply]
You need to use the same version of wasm-bindgen in the CLI and in the dependencies. You can search for `name = "wasm-bindgen"` in your Cargo.lock file. Then either try and update your dependency or try to set your CLI to the same version. As you just updated and it works on a fresh project, I'm guessing your game as an older version locked, so probably try to upgrade 👍
